pkg/awsiamauth/reconciler/reconciler.go

```diff
@@ -0,0 +1,161 @@
+package reconciler
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"github.com/google/uuid"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	anywherev1 "github.com/aws/eks-anywhere/pkg/api/v1alpha1"
+	"github.com/aws/eks-anywhere/pkg/awsiamauth"
+	anywhereCluster "github.com/aws/eks-anywhere/pkg/cluster"
+	"github.com/aws/eks-anywhere/pkg/constants"
+	"github.com/aws/eks-anywhere/pkg/controller"
+	"github.com/aws/eks-anywhere/pkg/controller/clientutil"
+	"github.com/aws/eks-anywhere/pkg/controller/clusters"
+	"github.com/aws/eks-anywhere/pkg/controller/serverside"
+	"github.com/aws/eks-anywhere/pkg/crypto"
+)
+
+// RemoteClientRegistry defines methods for remote cluster controller clients.
+type RemoteClientRegistry interface {
+	GetClient(ctx context.Context, cluster client.ObjectKey) (client.Client, error)
+}
+
+// Reconciler allows to reconcile AWSIamConfig.
+type Reconciler struct {
+	certgen              crypto.CertificateGenerator
+	templateBuilder      *awsiamauth.TemplateBuilder
+	generateUUID         UUIDGenerator
+	client               client.Client
+	remoteClientRegistry RemoteClientRegistry
+}
+
+// UUIDGenerator generates a new UUID.
+type UUIDGenerator func() uuid.UUID
+
+// New returns a new Reconciler.
+func New(certgen crypto.CertificateGenerator, generateUUID UUIDGenerator, client client.Client, remoteClientRegistry RemoteClientRegistry) *Reconciler {
+	return &Reconciler{
+		certgen:              certgen,
+		templateBuilder:      &awsiamauth.TemplateBuilder{},
+		generateUUID:         generateUUID,
+		client:               client,
+		remoteClientRegistry: remoteClientRegistry,
+	}
+}
+
+// EnsureCASecret ensures the AWS IAM Authenticator secret is present.
+// It uses a controller.Result to indicate when requeues are needed.
+func (r *Reconciler) EnsureCASecret(ctx context.Context, logger logr.Logger, cluster *anywherev1.Cluster) (controller.Result, error) {
+	clusterName := cluster.Name
+	s := &corev1.Secret{}
+	err := r.client.Get(ctx, types.NamespacedName{Name: awsiamauth.CASecretName(clusterName), Namespace: constants.EksaSystemNamespace}, s)
+	if apierrors.IsNotFound(err) {
+		logger.Info("Creating aws-iam-authenticator CA secret")
+		return r.createCASecret(ctx, cluster)
+	}
+	if err != nil {
+		return controller.Result{}, fmt.Errorf("fetching secret %s: %v", awsiamauth.CASecretName(clusterName), err)
+	}
+
+	logger.Info("aws-iam-authenticator CA secret found. Skipping secret create.", "Name", awsiamauth.CASecretName(clusterName))
```

Comment:
nit
```suggestion
	logger.Info("aws-iam-authenticator CA secret found. Skipping secret create.", "name", awsiamauth.CASecretName(clusterName))
```

Reply:
nit: can we store the secret name in a variable? so we don't need call this method twice